### PR TITLE
Bratseth/node metrics

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/ResourceIterator.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        this.memoryLoad = memoryLoad;
+        this.diskLoad = diskLoad;
+
+        // ceil: If the division does not produce a whole number we assume some node is missing
+        groupSize = (int)Math.ceil((double)currentAllocation.nodes() / currentAllocation.groups());
+        allocation = currentAllocation;
+
+        // What number of nodes is it effective to add or remove at the time from this cluster?
+        // This is the group size, since we (for now) assume the group size is decided by someone wiser than us
+        // and we decide tyhe number of groups.
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/ResourceIterator.java]:
```suggestion
        // and we decide the number of groups.
```

**Review by freva — APPROVED** [review]
LGTM to start testing, though resource calculation in AWS needs to be fixed before we actually start provisioning.

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsFetcherTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        orchestrator.suspend(new HostName("host-4.yahoo.com"));
+
+        {
+            httpClient.cannedResponse = cannedResponseForApplication1;
+            List<NodeMetrics.MetricValue> values = new ArrayList<>(fetcher.fetchMetrics(application1));
+            assertEquals("http://host-1.yahoo.com:4080/metrics/v2/values?consumer=vespa-consumer-metrics",
+                         httpClient.requestsReceived.get(0));
+            assertEquals(5, values.size());
+            assertEquals("metric value cpu.util: 16.2 at 1234 for host-1.yahoo.com", values.get(0).toString());
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsFetcherTest.java]: I think this is more readable. Afaik Float.toString() is locale-independent, which is one of the benefits of not using String.format in most cases ...

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsDb.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    /** Lock all access for now since we modify lists inside a map */
+    private final Object lock = new Object();
+
+    /** Add a measurement to this */
+    public void add(Collection<NodeMetrics.MetricValue> metricValues) {
+        synchronized (lock) {
+            for (var value : metricValues) {
+                List<Measurement> measurements = db.computeIfAbsent(new MeasurementKey(value.hostname(),
+                                                                                       Resource.fromMetric(value.name())),
+                                                                    (__) -> new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsDb.java]: I want an array and then it's just a question of whether the additional API in Deque is useful. I don't think so right now, but maybe later ...

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsDbTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        ManualClock clock = new ManualClock();
+        NodeMetricsDb db = new NodeMetricsDb();
+        List<NodeMetrics.MetricValue> values = new ArrayList<>();
+        for (int i = 0; i < 40; i++) {
+            values.add(new NodeMetrics.MetricValue("host0", "cpu.util", clock.instant().toEpochMilli(), 0.9f));
+            clock.advance(Duration.ofHours(1));
+        }
+        db.add(values);
+
+        assertEquals(32, db.getWindow(clock.instant().minus(Duration.ofHours(30)), Resource.cpu,    List.of("host0")).measurementCount());
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/autoscale/NodeMetricsDbTest.java]: You are right about the brackets, thanks! 
Stream version: or even just

```
                return (int) keys.stream()
                                 .flatMap(key -> db.getOrDefault(key, List.of()).stream())
                                 .filter(measurement -> measurement.timestamp >= startTime)
                                 .count();
```

I guess this is way more expensive though, but you are right it's too early to worry about it.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/Autoscaler.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                    return Optional.of(new ClusterResourcesWithCost(resources.with(nodeResources),
+                                                                    costOf(nodeResources) * resources.nodes()));
+            return Optional.empty();
+        }
+        else {
+            // return the cheapest flavor satisfying the target resources, if any
+            double bestCost = Double.MAX_VALUE;
+            Optional<Flavor> bestFlavor = Optional.empty();
+            for (Flavor flavor : nodeRepository.getAvailableFlavors().getFlavors()) {
+                if ( ! flavor.resources().satisfies(resources.nodeResources())) continue;
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/Autoscaler.java]: Thanks! I'll do this in another PR ...

**Pull request merged**